### PR TITLE
feat(frontend): Use Zod for Token type

### PR DESCRIPTION
**Pull request by AntonioVentilii-DFINITY**
# Motivation

We ultimately use `Zod` for `Token` type. That requires a few more sub-interfaces to use `Zod` too (for example the ones related to OnRamper).

### CODE REVIEW

File: src/frontend/src/lib/types/token.ts [hunk 1]
```diff
@@ -31,32 +31,41 @@ const TokenAppearanceSchema = z.object({
...
+});
+
+const TokenSchema = z
+	.object({
+		id: TokenIdSchema,
+		network: NetworkSchema,
+		standard: TokenStandardSchema,
+		category: TokenCategorySchema
+	})
+	.merge(TokenMetadataSchema)
```
Comment (peterpeterparker) [inline src/frontend/src/lib/types/token.ts:50]: Is `extend` closer to original definition?

**Review by peterpeterparker — APPROVED** [review]
Thx!

File: src/frontend/src/lib/types/token.ts (hunk 1 above)
Comment (AntonioVentilii-DFINITY) [reply src/frontend/src/lib/types/token.ts:50]: Unless I am mistaken (I haven't use `.merge` or `.extend` in Zod in a long time), to use `.extend`, I shall do

`.extend(TokenMetadataSchema.shape)`

Or not?

But if I look at `.merge` [documentation](https://zod.dev/?id=merge), it says that it is equivalent to it:

Screenshot 2024-10-31 at 14 22 51


What is the most common way of doing it? WDYT?

Comment (peterpeterparker) [reply src/frontend/src/lib/types/token.ts:50]: I don't think you need `.shape`, at least in the `oisy-wallet-signer` I just do extend.

https://github.com/dfinity/oisy-wallet-signer/blob/832a9d237109c99bb1fcec07e1b57c8457728eb3/src/types/rpc.ts#L36

But yeah if it's equivalent then all good.

**Review by peterpeterparker — APPROVED** [review]
LGTm, thx

File: src/frontend/src/lib/types/token.ts (hunk 1 above)
Comment (peterpeterparker) [reply src/frontend/src/lib/types/token.ts:50]: Ah maybe it's because in `oisy-wallet-signer` I extend with an object that contains the schema, gotcha.

**Pull request merged**